### ISSUE TITLE
GPU vs CPU Models in MnistDemo

### ISSUE THREAD

**Issue by fdncred** [cbovar/ConvNetSharp#89]
Any ideas why there is a difference between GPU and CPU models?

Here's my training code:
```csharp
private void MnistDemo()
{
    BuilderInstance.Volume = new VolumeBuilder();

    var datasets = new DataSets();
    if (!datasets.Load(100))
    {
        return;
    }

    // Create network
    _net = new Net<double>();
    _net.AddLayer(new InputLayer(28, 28, 1));
    _net.AddLayer(new ConvLayer(5, 5, 8) { Stride = 1, Pad = 2 });
    _net.AddLayer(new ReluLayer());
    _net.AddLayer(new PoolLayer(2, 2) { Stride = 2 });
    _net.AddLayer(new ConvLayer(5, 5, 16) { Stride = 1, Pad = 2 });
    _net.AddLayer(new ReluLayer());
    _net.AddLayer(new PoolLayer(3, 3) { Stride = 3 });
    _net.AddLayer(new FullyConnLayer(10));
    _net.AddLayer(new SoftmaxLayer(10));

    _trainer = new SgdTrainer(_net)
    {
        LearningRate = 0.01,
        BatchSize = 20,
        L2Decay = 0.001,
        Momentum = 0.9
    };

    var startTime = DateTime.Now;
    var epochStartTime = DateTime.Now;
    var totalTrainCount = datasets.Train.Count;
    var stepSize = _trainer.BatchSize;
    var numberOfTotalSteps = totalTrainCount / stepSize;

    Console.WriteLine("Convolutional neural network learning...[Press any key to stop]");
    //do
    //{
    for (int i = 0; i < numberOfTotalSteps; i++)
    {
        if (Console.KeyAvailable)
            break;

        epochStartTime = DateTime.Now;

        var trainSample = datasets.Train.NextBatch(_trainer.BatchSize);
        Train(trainSample.Item1, trainSample.Item2, trainSample.Item3);

        var testSample = datasets.Test.NextBatch(_trainer.BatchSize);
        Test(testSample.Item1, testSample.Item3, _testAccWindow);

        Console.WriteLine($"Loss: {_trainer.Loss} Train accuracy: {Math.Round(_trainAccWindow.Items.Average() * 100.0, 2)}% Test accuracy: {Math.Round(_testAccWindow.Items.Average() * 100.0, 2)}%");
        Console.WriteLine($"  Examples seen: {i} {_stepCount} of {totalTrainCount} Fwd: {Math.Round(_trainer.ForwardTimeMs, 2)}ms Bckw: {Math.Round(_trainer.BackwardTimeMs, 2)}ms");

        TimeSpan timeSpan = DateTime.Now - epochStartTime;
        int epochsRemainingByCount = totalTrainCount - _stepCount;
        double minutesRemaining = epochsRemainingByCount * timeSpan.TotalMinutes;
        DateTime finishTime = DateTime.Now + TimeSpan.FromMinutes(minutesRemaining);
        float percComplete = (float)((float)_stepCount / (float)totalTrainCount) * 100f;
        var timeTakenSoFar = DateTime.Now - startTime;
        Console.WriteLine($"    % Complete = {percComplete.ToString("N3")} % - {Math.Round(timeSpan.TotalMinutes)}m ({Math.Round(timeSpan.TotalSeconds)}s), eta {finishTime.ToShortTimeString()}, taken so far {timeTakenSoFar}");
        Console.WriteLine($"epochRemaining={epochsRemainingByCount} minRemaining={minutesRemaining} Now={DateTime.Now.ToShortTimeString()}");
    }
    //} while (!Console.KeyAvailable);

    var json = _net.ToJson();
    File.WriteAllText("net.model-full.json", json);
}
```
Here's the testing code:
```csharp
private void TestModel()
{
    var json = File.ReadAllText("net.model-full.json");
    var model = SerializationExtensions.FromJson<double>(json);
    var bmp = (Bitmap)Image.FromFile(@"C:\Users\Path\To\100.0.png");

    var vol = ConvertToVolume(bmp);
    var res = model.Forward(vol).Storage.ToArray();
    var prediction = model.GetPrediction();
}
```
For completeness here's the ConvertToVolume method:
```csharp
unsafe private Volume<double> ConvertToVolume(Bitmap bmp)
{
    //var vol = ConvNetSharp.Volume.GPU.Double.BuilderInstance.Volume.From(new double[bmp.Width * bmp.Height], new Shape(bmp.Width, bmp.Height, 1, 1));
    var vol = BuilderInstance.Volume.From(new double[bmp.Width * bmp.Height], new Shape(bmp.Width, bmp.Height, 1, 1));
    var data = bmp.LockBits(new Rectangle(0, 0, bmp.Width, bmp.Height), System.Drawing.Imaging.ImageLockMode.ReadOnly, bmp.PixelFormat);
    byte* src = (byte*)data.Scan0.ToPointer();
    var width = data.Width;
    var height = data.Height;
    int dst = 0;
    int pixelSize = Image.GetPixelFormatSize(bmp.PixelFormat) / 8;
    int offset = data.Stride - width * pixelSize;

    for (int y = 0; y < height; y++)
    {
        for (int x = 0; x < width; x++, dst++, src += pixelSize)
        {
            double pixel = *src / 255d;
            vol.Set(x, y, 0, 0, pixel);
        }
        src += offset;
    }
    bmp.UnlockBits(data);

    return vol;
}
```
Here are the two models, trained on the same data but give different results.
[net.model-full-CPU.json.txt](https://github.com/cbovar/ConvNetSharp/files/1775190/net.model-full-CPU.json.txt)
[net.model-full-GPU.json.txt](https://github.com/cbovar/ConvNetSharp/files/1775192/net.model-full-GPU.json.txt)

**Comment by cbovar** [comment]
Hi, What kind of difference do you see ? 
If you compare the two files, the weights will be different because they are initiated randomly.

**Comment by fdncred** [comment]
With this image
![0](https://user-images.githubusercontent.com/343840/36991766-5c0de3b2-206e-11e8-8fee-b88daebd1f0a.png)
And the CPU model, I get these results:
```
CPU
-	res	{double[10]}	double[]
		[0]	0.19452938990722496	double
		[1]	0.000796911776060907	double
		[2]	0.0034073427213554609	double
		[3]	0.016585338155885778	double
		[4]	0.0010524275880460284	double
		[5]	0.0010967844841814594	double
		[6]	0.263684351908287	double
		[7]	9.6973903125122763E-06	double
		[8]	0.51641612475875909	double
		[9]	0.0024216313098867608	double
prediction = 8
```
With the GPU model I get these results:
```
GPU
-	res	{double[10]}	double[]
		[0]	0.0880451832499159	double
		[1]	0.0087273514889211791	double
		[2]	0.011613476980292258	double
		[3]	0.0019632977126749951	double
		[4]	0.00041913240548843803	double
		[5]	0.00022204025327415854	double
		[6]	0.52718144755420915	double
		[7]	2.4428931258499982E-05	double
		[8]	0.36146206377997259	double
		[9]	0.0003415776439928138	double
prediction = 6
```

**Comment by cbovar** [comment]
Are you loading `net.model-full-CPU.json.txt` for the CPU version and `net.model-full-GPU.json.txt` for the GPU version?
Have you tried using the same model for CPU and GPU?

**Comment by cbovar** [comment]
I have tried to classify your image using model `net.model-full-CPU.json.txt` in CPU and GPU mode and I got the same prediction (8).

The serialized model is independant of the way it is computed: e.g. you can teach a model using GPU, serialize it, and then load it in CPU mode.

**Comment by fdncred** [comment]
To answer your first question, No, I have never tried using the CPU model during GPU prediction or vice-versa, but I can.

I understand what you're saying about the serialized model but perhaps I'm not making myself clear. Let me try again.

1. I expect to be able to train on CPU and use on CPU or GPU.
2. I expect to be able to train on GPU and use on CPU or GPU.
3. I expect to be able to train on CPU and train on GPU and the models be identical, or at least, produce the same results.
4. My post here with all the code is mean to give you or someone else a testable app to get consistent results. It's not doing that and I'm not sure if that's because of my ignorance, a bug in your code or a bug in my code.
5. I'm really just trying to figure out why the two models give different results, and more importantly, why the result is wrong. It should be returning 0 since I trained it on all 60k mnist data. I realize the model won't be 100% but this image is supposed to be an "easy" one to predict.

Thanks for trying to help,
Darren

**Comment by cbovar** [comment]
Understood. I believe MNIST has dark background and bright digits. Try to inverse the pixel brightness (or invert the picture directly):

` double pixel = 1.0 - *src / 255d;`

rather than

` double pixel = *src / 255d;`

**Comment by cbovar** [comment]
I get '0' with both models :)

**Comment by fdncred** [comment]
Ah... inversion. I get the same results too! Thanks for the help. Here's my detailed results
GPU Model
```
-		res	{double[10]}	double[]
		[0]	0.99984733078945909	double
		[1]	8.91605911494397E-09	double
		[2]	1.8327066007383681E-05	double
		[3]	1.043201565600329E-06	double
		[4]	1.2982100677518365E-09	double
		[5]	3.3125406848304891E-07	double
		[6]	2.7270446160438941E-05	double
		[7]	2.6661264175811447E-08	double
		[8]	8.9117934556281016E-05	double
		[9]	1.6542432649129405E-05	double
```
CPU Model
```
-		res	{double[10]}	double[]
		[0]	0.99996766686073535	double
		[1]	4.2000795875396795E-10	double
		[2]	1.0688763859241323E-06	double
		[3]	4.7782896901818784E-06	double
		[4]	3.92098033827561E-10	double
		[5]	4.8628629862657522E-07	double
		[6]	1.3847677483668293E-05	double
		[7]	3.8797388588320226E-09	double
		[8]	8.1440115134758284E-06	double
		[9]	4.0033060481202814E-06	double
```
While the results aren't identical (I assumed they would be) the key was to guess correctly with both models, which it did.

**Comment by fdncred** [comment]
Here's the same results without the scientific notation. It makes it easier to see that all the other guesses are so far off that they don't matter.

GPU Model
```
-		res	{double[10]}	double[]
		[0]	0.99984733078945909	double
		[1]	0.00000000891605911494397	double
		[2]	0.000018327066007383681	double
		[3]	0.000001043201565600329	double
		[4]	0.0000000012982100677518365	double
		[5]	0.00000033125406848304891	double
		[6]	0.000027270446160438941	double
		[7]	0.000000026661264175811447	double
		[8]	0.000089117934556281016	double
		[9]	0.000016542432649129405	double
```
CPU Model
```
-		res	{double[10]}	double[]
		[0]	0.99996766686073535	double
		[1]	0.00000000042000795875396795	double
		[2]	0.0000010688763859241323	double
		[3]	0.0000047782896901818784	double
		[4]	0.000000000392098033827561	double
		[5]	0.00000048628629862657522	double
		[6]	0.000013847677483668293	double
		[7]	0.0000000038797388588320226	double
		[8]	0.0000081440115134758284	double
		[9]	0.0000040033060481202814	double
```
